runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/JobGroupTest.java

```diff
@@ -1293,6 +1313,80 @@ public IStatus run(IProgressMonitor monitor) {
 		assertTrue("2.0", numShouldCancelCalled[0] < NUM_JOBS_LIMIT + 10);
 	}
 
+	/**
+	 * Tested scenario: - 2 jobs are started in parallel and the first one returns a
+	 * CANCEL_STATUS while the second one is still running. This lets
+	 * <code>JobGroup.shouldCancel</code> return <code>true</code> which triggers
+	 * the cancelation of any running job.
+	 *
+	 * Expected result: The second job is explicitly canceled
+	 *
+	 */
+	public void testShouldCancel_5() {
+		final CountDownLatch readyToRun = new CountDownLatch(2);
+		final CountDownLatch startRunning = new CountDownLatch(1);
+		final CountDownLatch explicitlyCanceled = new CountDownLatch(1);
+
+		// the job group allows for 2 threads so both jobs will be started in
+		// parallel
+		final JobGroup jobGroup = new JobGroup("JobGroup", 2, 2) {
+			@Override
+			protected boolean shouldCancel(IStatus lastCompletedJobResult, int numberOfFailedJobs,
+					int numberOfCanceledJobs) {
+				// cancel the group as soon as one job in it returns the CANCEL_STATUS
+				return numberOfCanceledJobs > 0;
+			}
+		};
+
+		Job returnsCancelStatusJob = Job.create("The one that cancels", __ -> {
+			readyToRun.countDown();
+			awaitLatch(startRunning);
+
+			return Status.CANCEL_STATUS;
+		});
+
+		Job canceledJob = new Job("The one that is explicitly canceled") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				readyToRun.countDown();
+				awaitLatch(startRunning);
+
+				// give it time: this Job needs to be running at the time the other job finishes
+				awaitLatch(explicitlyCanceled);
+				return Status.OK_STATUS;
+			}
+
+			@Override
+			protected void canceling() {
+				explicitlyCanceled.countDown();
+			}
+		};
```

Comment:
I propose the usage of a barrier instead of multiple latches as follows for these reasons:
1. There is no necessity to synchronize the jobs via the main thread, they can simply trigger each other.
2. The barrier already provides the functionality (and more) as defined in `awaitLatch()`.
3. The barrier also prints a stack dump in case the test does not finish. This is important for debugging why one of the jobs did not proceed.

With the proposed change, you can replace all the latch logic in the main thread with a `testBarrier.waitForStatus(TestBarrier2.STATUS_DONE)` that prints a stack trace in case the job waiting for cancellation does not finish.


```suggestion
		Job returnsCancelStatusJob = Job.create("The one that cancels", __ -> {
			testBarrier.setStatus(TestBarrier2.STATUS_START);
			testBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
			return Status.CANCEL_STATUS;
		});

		Job canceledJob = new Job("The one that is explicitly canceled") {
			@Override
			protected IStatus run(IProgressMonitor monitor) {
				testBarrier.waitForStatus(TestBarrier2.STATUS_START);
				testBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
				testBarrier.waitForStatus(TestBarrier2.STATUS_DONE);
				return Status.OK_STATUS;
			}

			@Override
			protected void canceling() {
				testBarrier.setStatus(TestBarrier2.STATUS_DONE);
			}
		};
```

Reply:
Thank you for the suggestion. I changed the code but used 2 barriers instead of one. I find it easier to read if every job has its own barrier and they are both orchestrated from the main thread.